src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp

```diff
@@ -426,68 +424,29 @@ void ShenandoahBarrierSet::arraycopy_barrier(T* src, T* dst, size_t count) {
     return;
   }
 
-  char gc_state = ShenandoahThreadLocalData::gc_state(Thread::current());
+  const char gc_state = ShenandoahThreadLocalData::gc_state(Thread::current());
+  if ((gc_state & ShenandoahHeap::MARKING) != 0) {
```

Comment:
A question on sequencing. Can we have marking and update-refs at the same time? If so, should update-refs happen first, and then marking should be doing the walk over (fixed-up) objects?